lib/Firtool/Firtool.cpp

```diff
@@ -59,16 +59,15 @@ LogicalResult firtool::populateCHIRRTLToLowFIRRTL(mlir::PassManager &pm,
 
   pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInjectDUTHierarchyPass());
 
+  if (!opt.shouldDisableOptimization())
+    pm.nest<firrtl::CircuitOp>().nestAny().addPass(mlir::createCSEPass());
+
   pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
       firrtl::createPassiveWiresPass());
 
   pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
       firrtl::createDropNamesPass(opt.getPreserveMode()));
 
-  if (!opt.shouldDisableOptimization())
-    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
-        mlir::createCSEPass());
```

Comment:
Is there any benefit in running twice or keeping this here to run on `FModuleOp`? Or: is CSE blocked on things with different names / does `DropNames` expose work for CSE?

Reply:
Should be OK I think.  CSE will only remove the ops marked Pure, and anything FNamable is not Pure.